docs/query-with-spark-sql-performacne-tuning.md

```diff
@@ -0,0 +1,58 @@
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and 
+    limitations under the License.
+-->
+
+# Query with spark-sql performacne tuning
+  This tutorial guides you to create CarbonData Tables and optimize performance.
+  The following sections will elaborate on the below topics :
+
+  * [The influence of spark.sql.codegen.wholeStage configuration on query](#The influence of spark.sql.codegen.wholeStage configuration on query)
+
+## The influence of spark.sql.codegen.wholeStage configuration on query 
+
+In practice, we found that when the sum of CarbonData's queries reaches a certain threshold, the query time increases dramatically. As shown in the figure below(spark 2.1):
+
+![File Directory Structure](../docs/images/codegen.png?raw=true)
+
+The horizontal axis is the number of sum, and the vertical axis is the time consumed in seconds.
+
+It can be seen from the figure that when the number of sum exceeds 85, the query time is significantly increased.
+
+After analysis, this problem is related to spark.sql.codegen.wholeStage, which is enabled by default for spark 2.0. and it will do all the *internal optimization possible from the spark catalist side*. [https://jaceklaskowski.gitbooks.io/mastering-spark-sql/spark-sql-whole-stage-codegen.html](https://jaceklaskowski.gitbooks.io/mastering-spark-sql/spark-sql-whole-stage-codegen.html)
+
+**Whole-Stage Java Code Generation** (aka *Whole-Stage CodeGen*) is a physical query optimization in Spark SQL that fuses multiple physical operators (as a subtree of plans that [support code generation](https://jaceklaskowski.gitbooks.io/mastering-spark-sql/spark-sql-CodegenSupport.html)) together into a single Java function.
+
+Whole-Stage Java Code Generation improves the execution performance of a query by collapsing a query tree into a single optimized function that eliminates virtual function calls and leverages CPU registers for intermediate data.
+
+When sum is too large, the logic calculation function of hashaggregate operator is too large (there are nearly 3000 lines of code when there are 34 indicators). Java itself does not recommend too large methods, which will reduce the processing efficiency, and exceed the JIT threshold, making the final execution in the way of interpretation. When the carbon grows in the counter, the performance drops sharply.
+
+But unfortunately, spark 2.1 only provides switching capability. User can only choose to turn the function on or off. This leads to a sharp drop in performance when the aggregation operator is too large.
+
+Fortunately, spark 2.3 provide more configuration. Users can better configure this parameter. 
+
+So in spark 2.1, when the number of operators cannot be confirmed, spark.sql.codegen.wholeStage can be turned off to ensure the query efficiency. when in spark 2.3 and above users can open spark.sql.codegen.wholeStage  and configure it.  
```

Comment:
Line 57 also talking about configuration for spark 2.1, but the meaning is different. Better to unify them.

Reply:
It's reasonable. I'll adjust it.